File: windows/src/FileMCP.Core/ToolExecutionContext.cs
using System.Text.Json.Nodes;

namespace FileMCP.Core;

internal sealed record ToolBudgetLimits(
    int MaxVisitedEntries,
    int MaxFilesScanned,
    long MaxBytesScanned,
    int MaxOutputItems,
    int TimeoutMs)
{
    public static readonly ToolBudgetLimits ServerCaps = new(
        FileMcpConstants.MaxSearchVisited,
        FileMcpConstants.MaxSearchVisited,
        FileMcpConstants.MaxSearchContentBytesScanned,
        FileMcpConstants.MaxListEntries,
        120_000);
}

internal sealed class ToolExecutionContext : IDisposable
{
    public const string BudgetMetadataKey = "io.filemcp/budget";

    private readonly CancellationToken _parentToken;
    private readonly Func<bool> _cancellationProbe;
    private readonly CancellationTokenSource _deadlineSource;
    private int _visitedEntries;
    private int _filesScanned;
    private long _bytesScanned;
    private int _outputItems;
    private bool _truncated;
    private string _truncationReason = "none";

    private ToolExecutionContext(ToolBudgetLimits limits, CancellationToken parent, Func<bool>? cancellationProbe)
    {
        Limits = limits;
        _parentToken = parent;
        _cancellationProbe = cancellationProbe ?? (() => false);
        _deadlineSource = CancellationTokenSource.CreateLinkedTokenSource(parent);
        _deadlineSource.CancelAfter(TimeSpan.FromMilliseconds(limits.TimeoutMs));
    }

    public ToolBudgetLimits Limits { get; }
    public CancellationToken CancellationToken => _deadlineSource.Token;
    public bool Truncated => _truncated;
    public string TruncationReason => _truncationReason;
    public int RemainingFiles => Math.Max(0, Limits.MaxFilesScanned - _filesScanned);
    public long RemainingBytes => Math.Max(0, Limits.MaxBytesScanned - _bytesScanned);

    public static bool HasBudget(JsonObject? meta) => meta?[BudgetMetadataKey] is not null;

    public static ToolExecutionContext Create(JsonObject? meta, CancellationToken parent = default, Func<bool>? cancellationProbe = null)
    {
        var limits = ToolBudgetLimits.ServerCaps;
        if (meta?[BudgetMetadataKey] is JsonNode rawBudget)
        {
            if (rawBudget is not JsonObject budget)
                throw new FileMcpException($"{BudgetMetadataKey} must be an object");
            var allowed = new HashSet<string>(StringComparer.Ordinal)
            {
                "maxVisitedEntries", "maxFilesScanned", "maxBytesScanned", "maxOutputItems", "timeoutMs",
            };
            foreach (var pair in budget)
                if (!allowed.Contains(pair.Key)) throw new FileMcpException($"Unknown budget field: {pair.Key}");

            limits = new ToolBudgetLimits(
                LowerInt(budget, "maxVisitedEntries", limits.MaxVisitedEntries),
                LowerInt(budget, "maxFilesScanned", limits.MaxFilesScanned),
                LowerLong(budget, "maxBytesScanned", limits.MaxBytesScanned),
                LowerInt(budget, "maxOutputItems", limits.MaxOutputItems),
                LowerInt(budget, "timeoutMs", limits.TimeoutMs));
        }
        return new ToolExecutionContext(limits, parent, cancellationProbe);
    }

    public bool TryContinue()
    {
        if (_cancellationProbe()) return Truncate("cancelled");
        if (!CancellationToken.IsCancellationRequested) return true;
        return Truncate(_parentToken.IsCancellationRequested ? "cancelled" : "timeout");
    }

    public bool TryVisitEntry()
    {
        if (!TryContinue()) return false;
        if (_visitedEntries >= Limits.MaxVisitedEntries) return Truncate("visited_entries");
        _visitedEntries++;
        return true;
    }

    public bool TryScanFile(long bytes)
    {
        if (!TryContinue()) return false;
        if (_filesScanned >= Limits.MaxFilesScanned) return Truncate("files_scanned");
        if (bytes < 0 || bytes > Limits.MaxBytesScanned - _bytesScanned) return Truncate("bytes_scanned");
        _filesScanned++;
        _bytesScanned += bytes;
        return true;
    }

    public bool TryOutputItem()
    {
        if (!TryContinue()) return false;
        if (_outputItems >= Limits.MaxOutputItems) return Truncate("output_items");
        _outputItems++;
        return true;
    }

    public void MarkTruncated(string reason) => Truncate(reason);

    public JsonObject Usage() => new()
    {
        ["visitedEntries"] = _visitedEntries,
        ["filesScanned"] = _filesScanned,
        ["bytesScanned"] = _bytesScanned,
        ["outputItems"] = _outputItems,
        ["truncated"] = _truncated,
        ["truncationReason"] = _truncationReason,
        ["budget"] = new JsonObject
        {
            ["maxVisitedEntries"] = Limits.MaxVisitedEntries,
            ["maxFilesScanned"] = Limits.MaxFilesScanned,
            ["maxBytesScanned"] = Limits.MaxBytesScanned,
            ["maxOutputItems"] = Limits.MaxOutputItems,
            ["timeoutMs"] = Limits.TimeoutMs,
        },
    };

    public void Dispose() => _deadlineSource.Dispose();

    private bool Truncate(string reason)
    {
        _truncated = true;
        if (_truncationReason == "none") _truncationReason = string.IsNullOrWhiteSpace(reason) ? "unknown" : reason;
        return false;
    }

    private static int LowerInt(JsonObject budget, string key, int serverCap)
    {
        if (budget[key] is null) return serverCap;
        if (budget[key] is not JsonValue value || !value.TryGetValue<int>(out var requested) || requested <= 0)
            throw new FileMcpException($"Budget field {key} must be a positive integer");
        if (requested > serverCap) throw new FileMcpException($"Budget field {key} may only lower the server cap ({serverCap})");
        return requested;
    }

    private static long LowerLong(JsonObject budget, string key, long serverCap)
    {
        if (budget[key] is null) return serverCap;
        if (budget[key] is not JsonValue value || !value.TryGetValue<long>(out var requested) || requested <= 0)
            throw new FileMcpException($"Budget field {key} must be a positive integer");
        if (requested > serverCap) throw new FileMcpException($"Budget field {key} may only lower the server cap ({serverCap})");
        return requested;
    }
}

File: macos/ToolExecutionContext.swift
import Foundation

enum ToolExecutionContextError: LocalizedError {
    case invalid(String)
    var errorDescription: String? {
        switch self { case let .invalid(message): return message }
    }
}

struct ToolBudgetLimits {
    let maxVisitedEntries: Int
    let maxFilesScanned: Int
    let maxBytesScanned: Int64
    let maxOutputItems: Int
    let timeoutMs: Int

    static let serverCaps = ToolBudgetLimits(
        maxVisitedEntries: 50_000,
        maxFilesScanned: 50_000,
        maxBytesScanned: 50_000_000,
        maxOutputItems: 1_000,
        timeoutMs: 120_000
    )
}

final class ToolExecutionContext {
    static let budgetMetadataKey = "io.filemcp/budget"

    let limits: ToolBudgetLimits
    private let nowProvider: () -> Date
    private let cancellationProbe: () -> Bool
    private let deadline: Date

    private(set) var visitedEntries = 0
    private(set) var filesScanned = 0
    private(set) var bytesScanned: Int64 = 0
    private(set) var outputItems = 0
    private(set) var truncated = false
    private(set) var truncationReason = "none"
    var remainingFiles: Int { max(0, limits.maxFilesScanned - filesScanned) }
    var remainingBytes: Int64 { max(0, limits.maxBytesScanned - bytesScanned) }

    init(
        meta: [String: Any]?,
        nowProvider: @escaping () -> Date = Date.init,
        cancellationProbe: @escaping () -> Bool = { false }
    ) throws {
        var limits = ToolBudgetLimits.serverCaps
        if let rawBudget = meta?[Self.budgetMetadataKey] {
            guard let budget = rawBudget as? [String: Any] else {
                throw ToolExecutionContextError.invalid("\(Self.budgetMetadataKey) must be an object")
            }
            let allowed: Set<String> = [
                "maxVisitedEntries", "maxFilesScanned", "maxBytesScanned", "maxOutputItems", "timeoutMs",
            ]
            for key in budget.keys where !allowed.contains(key) {
                throw ToolExecutionContextError.invalid("Unknown budget field: \(key)")
            }
            limits = ToolBudgetLimits(
                maxVisitedEntries: try Self.lowerInt(budget, "maxVisitedEntries", limits.maxVisitedEntries),
                maxFilesScanned: try Self.lowerInt(budget, "maxFilesScanned", limits.maxFilesScanned),
                maxBytesScanned: try Self.lowerInt64(budget, "maxBytesScanned", limits.maxBytesScanned),
                maxOutputItems: try Self.lowerInt(budget, "maxOutputItems", limits.maxOutputItems),
                timeoutMs: try Self.lowerInt(budget, "timeoutMs", limits.timeoutMs)
            )
        }

        self.limits = limits
        self.nowProvider = nowProvider
        self.cancellationProbe = cancellationProbe
        self.deadline = nowProvider().addingTimeInterval(Double(limits.timeoutMs) / 1000.0)
    }

    static func hasBudget(_ meta: [String: Any]?) -> Bool {
        meta?[budgetMetadataKey] != nil
    }

    @discardableResult
    func tryContinue() -> Bool {
        if cancellationProbe() { return truncate("cancelled") }
        if nowProvider() >= deadline { return truncate("timeout") }
        return true
    }

    @discardableResult
    func tryVisitEntry() -> Bool {
        guard tryContinue() else { return false }
        guard visitedEntries < limits.maxVisitedEntries else { return truncate("visited_entries") }
        visitedEntries += 1
        return true
    }

    @discardableResult
    func tryScanFile(bytes: Int64) -> Bool {
        guard tryContinue() else { return false }
        guard filesScanned < limits.maxFilesScanned else { return truncate("files_scanned") }
        guard bytes >= 0, bytes <= limits.maxBytesScanned - bytesScanned else { return truncate("bytes_scanned") }
        filesScanned += 1
        bytesScanned += bytes
        return true
    }

    @discardableResult
    func tryOutputItem() -> Bool {
        guard tryContinue() else { return false }
        guard outputItems < limits.maxOutputItems else { return truncate("output_items") }
        outputItems += 1
        return true
    }

    func markTruncated(_ reason: String) {
        _ = truncate(reason)
    }

    func usage() -> [String: Any] {
        [
            "visitedEntries": visitedEntries,
            "filesScanned": filesScanned,
            "bytesScanned": bytesScanned,
            "outputItems": outputItems,
            "truncated": truncated,
            "truncationReason": truncationReason,
            "budget": [
                "maxVisitedEntries": limits.maxVisitedEntries,
                "maxFilesScanned": limits.maxFilesScanned,
                "maxBytesScanned": limits.maxBytesScanned,
                "maxOutputItems": limits.maxOutputItems,
                "timeoutMs": limits.timeoutMs,
            ],
        ]
    }

    @discardableResult
    private func truncate(_ reason: String) -> Bool {
        truncated = true
        if truncationReason == "none" {
            let trimmed = reason.trimmingCharacters(in: .whitespacesAndNewlines)
            truncationReason = trimmed.isEmpty ? "unknown" : trimmed
        }
        return false
    }

    private static func lowerInt(_ budget: [String: Any], _ key: String, _ serverCap: Int) throws -> Int {
        guard let raw = budget[key] else { return serverCap }
        guard !(raw is Bool), let number = raw as? NSNumber else {
            throw ToolExecutionContextError.invalid("Budget field \(key) must be a positive integer")
        }
        let value = number.intValue
        guard value > 0, number.doubleValue == Double(value) else {
            throw ToolExecutionContextError.invalid("Budget field \(key) must be a positive integer")
        }
        guard value <= serverCap else {
            throw ToolExecutionContextError.invalid("Budget field \(key) may only lower the server cap (\(serverCap))")
        }
        return value
    }

    private static func lowerInt64(_ budget: [String: Any], _ key: String, _ serverCap: Int64) throws -> Int64 {
        guard let raw = budget[key] else { return serverCap }
        guard !(raw is Bool), let number = raw as? NSNumber else {
            throw ToolExecutionContextError.invalid("Budget field \(key) must be a positive integer")
        }
        let value = number.int64Value
        guard value > 0, number.doubleValue == Double(value) else {
            throw ToolExecutionContextError.invalid("Budget field \(key) must be a positive integer")
        }
        guard value <= serverCap else {
            throw ToolExecutionContextError.invalid("Budget field \(key) may only lower the server cap (\(serverCap))")
        }
        return value
    }
}
